pkg/executor/statefulset.go

```diff
@@ -149,23 +149,71 @@ func UpdateStatefulSetNonReplicas(
 	// need/expect to be under our control, other than the replicas count,
 	// correct them here.
 
-	// For now only checking the owner reference.
-	if shared.OwnerReferencesPresent(cr, statefulSet.OwnerReferences) {
+	patchedRes := *statefulSet
+	needPatch := false
+
+	// Check the owner reference.
+	if !shared.OwnerReferencesPresent(cr, statefulSet.OwnerReferences) {
+		shared.LogInfof(
+			reqLogger,
+			cr,
+			shared.EventReasonNoEvent,
+			"repairing owner ref on statefulset{%s}",
+			statefulSet.Name,
+		)
+
+		// So, what to do. Do we add our owner ref to the existing ones? What if
+		// something else is claiming to be controller? Probably some stale ref
+		// left by a bad backup/restore process? We're just going to nuke any
+		// existing owner refs.
+		patchedRes.OwnerReferences = shared.OwnerReferences(cr)
+		needPatch = true
+	}
+
+	// EZML-862
+	// We should also compare the current statefulset container image against
+	// the image defined for the corresponding role in the KDApp spec. If images
+	// are not the same, KDApp spec image has a higher priority, so we shoud
+	// upgrade the statefulset one.
+	// Make sure, that according this logic, there is no sence to edit a statefulset
+	// directly, as it will be reconciled back to the KDApp spec state.
+	appRoleImage, err := catalog.ImageForRole(cr, role.Name)
+	if err != nil {
+		return err
+	}
+
+	containers := shared.StatefulSetContainers(statefulSet)
+	currentImage := containers[0].Image
+
+	if strings.Compare(appRoleImage, currentImage) != 0 {
+		patchedRes.Spec.Template.Spec.Containers = make([]v1.Container, len(containers))
+		patchedContainers := shared.StatefulSetContainers(&patchedRes)
+		copy(patchedContainers, containers)
+		patchedContainers[0].Image = appRoleImage
+
+		// EZML-865
+		// Set current role members count should be upgraded
+		// It will be used at the syncMembers() step
+		rs, err := shared.GetRoleStatusByName(cr, role.Name)
+		if err != nil {
+			return err
+		}
+
+		if (*rs).UpgradingMembers == nil {
+			(*rs).UpgradingMembers = make(map[string]*string)
+		}
```

Comment:
Thinking about the implications of the situation where UpgradingMembers already has some elements populated here. That would be bad, as I don't think we should be upgrading a role while a previous upgrade to that role is still cranking along.

In the current state of the live-app-upgrade branch, you have it so that this image reconciliation stuff will be skipped if the cluster is not in a ready state. That's good! But the cluster's overall state doesn't reflect whether some individual role is upgrading, so it won't protect us here.

We need to think/talk about the correct approach. Some possibilities:

- Skip doing the image reconciliation for this role if UpgradingMembers is non-empty.
- In the validator, reject changing the role image if that role's UpgradingMembers is non-empty.
- ??? probably other things we could do also.

At first glance I'm leaning toward that validator check as the correct solution.

Reply:
So, should we reject a role changing for all clusters because of some member of some cluster is in the upgrading state?

Reply:
Hmm.

Yeah that does sound like overkill. I guess the validator is not the best approach.

Ignoring the new image until the old image upgrade is finished, that may be the way to go unless some other solution comes to mind. Basically, we could say that UpgradingMembers will be populated here with the entries for this appRoleImage only if UpgradingMembers is currently empty.

Reply:
Well, I rejected image reconciliation on the statefulset-patching stage. But there appeared the problem, when user may edit wrong image tag or the new image may be unavailable by some reason. It causes the cluster falling into non-ready state + the target role is sealed that it is in upgrading status. The failed pod will endless restart and fail (ErrImagePull), so it may be healed by upgrading or rolling back app spec image again. There are few option how it could be resolved:

- Memorize last working version  and force rolling back the image to it.
- Add ability to change app spec and patch the statefulset again, despite the cluster is not ready.

Currently I chose the second case, as it is IMO more common and user may decide, if he wants just discard appspec changes or change wrong image to some another one. For it I added specific logic, contains:
- Added to `MemberStatus` struct the new string field `ContainerImage` that stores the last succesfully performed image for this concrete pod.
- Before patching a statefulset check, if its role status is in upgrading state and cluster is not ready in the same time. If so, check, if all role member (pods) are marked for upgrade (`UpgradingMembers` map) by the same image as defined in the app spec for this role. If so, do nothing, pod just could updates slowly. Otherwise, (appSpec image is different) it means user changed the role again. In this case we need to full clean the role `UpgradingMembers` map and mark that we need to restart the upgrade.
- Allow statefulset patching by `clusterIsReady && !roleIsUpgrading` flags combination or, alternatively, by `restartUpgrade` flag.
- Note, that if user decide to rollback to the last succesfully run image, it could be still run in the few members of the current role, except the failed one. In this case, we shouldn't add them to role upgrading members, as in this pods will not restart but the role will be always marked as upgrading and there will be not possible to upgrade it in the future. For this purpose we need to compare the `appSpecImage` with `MemberStatus.ContainerImage` field.